File: src/main/java/org/embulk/test/EmbulkPluginTest.java
package org.embulk.test;

import org.embulk.EmbulkEmbed;
import org.embulk.config.ConfigDiff;
import org.embulk.config.ConfigSource;
import org.embulk.exec.ResumeState;
import org.junit.Before;

public abstract class EmbulkPluginTest {
    private ExtendedTestingEmbulk embulk;

    @Before
    public void setup() {
        TestingEmbulk.Builder builder = ExtendedTestingEmbulk.builder();
        setup(builder);
        embulk = (ExtendedTestingEmbulk) builder.build();
    }

    protected void setup(TestingEmbulk.Builder builder) {
        // You can override this method in your test class
    }

    protected TestingEmbulk.RunResult runInput(ConfigSource inConfig) {
        return runInput(inConfig, null);
    }

    protected TestingEmbulk.RunResult runInput(ConfigSource inConfig, ConfigDiff confDiff) {
        return embulk.new RunConfig()
                .inConfig(inConfig)
                .configDiff(confDiff)
                .execConfig(newConfig().set("min_output_tasks", 1))
                .outConfig(newConfig().set("type", "test"))
                .run();
    }

    protected EmbulkEmbed.ResumableResult resume(ConfigSource inConfig) {
        return resume(inConfig, null);
    }

    protected EmbulkEmbed.ResumableResult resume(ConfigSource inConfig, ResumeState resumeState) {
        return embulk.new RunConfig()
                .inConfig(inConfig)
                .resumeState(resumeState)
                .execConfig(newConfig().set("min_output_tasks", 1))
                .outConfig(newConfig().set("type", "test"))
                .resume();
    }

    protected TestingEmbulk.RunResult runFilter(ConfigSource filterConfig, String inConfigPath) {
        return embulk.new RunConfig()
                .inConfig(ExtendedEmbulkTests.configFromResource(inConfigPath))
                .filterConfig(filterConfig)
                .execConfig(newConfig().set("min_output_tasks", 1))
                .outConfig(newConfig().set("type", "test"))
                .run();
    }

    protected ConfigSource newConfig() {
        return embulk.newConfig();
    }
}

File: src/main/kotlin/org/embulk/test/EmbulkPluginTest.kt
package org.embulk.test

import org.embulk.config.ConfigSource

abstract class EmbulkPluginTest {
    @Embulk
    internal lateinit var embulk: ExtendedTestingEmbulk

    protected fun runConfig(inConfigPath: String): ExtendedTestingEmbulk.RunConfig {
        return runConfig(configFromResource(inConfigPath))
    }

    protected fun runConfig(inConfig: ConfigSource): ExtendedTestingEmbulk.RunConfig {
        return embulk.RunConfig()
                .inConfig(inConfig)
                .execConfig(config().set("min_output_tasks", 1))
                .outConfig(config().set("type", "local_object"))
    }

    protected fun config(): ConfigSource {
        return embulk.newConfig()
    }

    protected fun <T> getInstance(type: Class<T>): T {
        return embulk.superEmbed.injector.getInstance(type)
    }

    protected fun setSystemConfig(systemConfig: ConfigSource) {
        embulk.setSystemConfig(systemConfig)
    }
}
